### PR TITLE
[FEAT] Include file paths as column from read_parquet/csv/json

### PR DESCRIPTION
**Pull request by colin-ho**
Addresses: https://github.com/Eventual-Inc/Daft/issues/2808

This PR enables adding file path as a column from file reads via the `file_path_column: str | None` parameter. This works by appending a column of the file path to the `Table` post read + pushdowns.

Having it as a string makes it easy to have unique field name guarantees, i.e. if the user specifies a column name that already exists then an error is thrown.

### CODE REVIEW

File: daft/io/_parquet.py [hunk 1]
```diff
@@ -45,9 +46,10 @@ def read_parquet(
         infer_schema (bool): Whether to infer the schema of the Parquet, defaults to True.
         schema (dict[str, DataType]): A schema that is used as the definitive schema for the Parquet file if infer_schema is False, otherwise it is used as a schema hint that is applied after the schema is inferred.
         io_config (IOConfig): Config to be used with the native downloader
+        file_path_column: Include the source path(s) as a column with this name. Defaults to None.
         use_native_downloader: Whether to use the native downloader instead of PyArrow for reading Parquet.
         coerce_int96_timestamp_unit: TimeUnit to coerce Int96 TimeStamps to. e.g.: [ns, us, ms], Defaults to None.
-        _multithreaded_io: Whether to use multithreading for IO threads. Setting this to False can be helpful in reducing
+        _multithreaded_io: Include the source path(s) as a column called
```
Comment (universalmind303) [inline daft/io/_parquet.py:52]: potential typo

Comment (colin-ho) [reply daft/io/_parquet.py:52]: Oops thank you

**Comment by kindkindk** [comment]
This is a good feature. To read file user will add `df.read_parquet("...", filename="True")`, note the string "True" and reason is explained in top comment. 

Can we instead have fileName as boolean in Python API and then Rust can treat it as string ?

**Comment by colin-ho** [comment]
> This is a good feature. To read file user will add `df.read_parquet("...", filename="True")`, note the string "True" and reason is explained in top comment.
> 
> Can we instead have fileName as boolean in Python API and then Rust can treat it as string ?

Thanks for the comment! and sorry for late reply. 

We could have it as a boolean in the Python API as you mentioned, but I don't quite understand what you mean by having Rust treat it as string. If it was a boolean, then on the Rust side we would simply create a column called "paths" for the file paths. The issue arises when the data _already_ has a column called "paths". 

In which case i can think of two solutions:
1. retry with different variations like "_paths", "file_paths", etc
2. expose two arguments (`include_file_path: bool = False` and `file_path_column_name: str | None = None`), in which we would use the provided `file_path_column_name` otherwise default to "paths", and subsequently if there is an existing column with that name, throw an error and tell the user to pick a different name. This solution is essentially a more explicit variant of the current implementation, which is just the single `file_path_column_name: str | None = None` argument.

cc @jaychia for thoughts

File: src/daft-plan/src/builder.rs [hunk 2]
```diff
@@ -658,6 +683,7 @@ impl ParquetScanBuilder {
             ))),
             self.infer_schema,
             self.schema,
+            None,
```
Comment (desmondcheongzx) [inline src/daft-plan/src/builder.rs:686]: I think here we should either add a TODO or add `file_path_column` to the `ParquetScanBuilder`, then add the argument to the SQL side as well in `read_parquet.rs`.

File: src/daft-scan/src/lib.rs [hunk 3]
```diff
@@ -451,6 +462,12 @@ impl ScanTask {
                 p2: sc2.pushdowns.clone(),
             });
         }
+        if sc1.file_path_column != sc2.file_path_column {
+            return Err(Error::DifferingFileFormatConfigsInScanTaskMerge {
```
Comment (desmondcheongzx) [inline src/daft-scan/src/lib.rs:466]:
```suggestion
            return Err(Error::DifferingFilePathColumnsInScanTaskMerge {
```

**Review by desmondcheongzx — APPROVED** [review]
Two comments but overall looks good!

File: src/daft-scan/src/lib.rs (hunk 3 above)
Comment (desmondcheongzx) [reply src/daft-scan/src/lib.rs]: oops my bad, didn't realize the argument names were different. CI is failing on that :\

Comment (colin-ho) [reply src/daft-scan/src/lib.rs]: yeah no worries, fixing

**Pull request merged**